forwardingproxy.go

```diff
@@ -0,0 +1,224 @@
+		AuthUser:            "",
+		AuthPass:            "",
+		DestDialTimeout:     10 * time.Second,
+		DestReadTimeout:     5 * time.Second,
+		DestWriteTimeout:    5 * time.Second,
+		ClientReadTimeout:   5 * time.Second,
+		ClientWriteTimeout:  5 * time.Second,
+	}
+}
+
+// Proxy is a ProxyFunc.
+func (p *ForwardingProxy) Proxy(w io.Writer, r io.ReadCloser, msg *proto.ControlMessage) {
+	switch msg.ForwardedProto {
+	case proto.HTTPCONNECT:
+		// ok
+	default:
+		p.logger.Log(
+			"level", 0,
+			"msg", "unsupported protocol",
+			"ctrlMsg", msg,
+		)
+		return
+	}
+
+	rw, ok := w.(http.ResponseWriter)
+	if !ok {
+		p.logger.Log(
+			"level", 0,
+			"msg", "expected http.ResponseWriter",
+			"ctrlMsg", msg,
+		)
+	}
+
+	req, err := http.ReadRequest(bufio.NewReader(r))
+	if err != nil {
+		p.logger.Log(
+			"level", 0,
+			"msg", "failed to read request",
+			"ctrlMsg", msg,
+			"err", err,
+		)
+		return
+	}
+	setXForwardedFor(req.Header, msg.RemoteAddr)
+	req.URL.Host = msg.ForwardedHost
+
+	p.ServeHTTP(rw, req, r)
+
+}
+
+func (p *ForwardingProxy) ServeHTTP(w http.ResponseWriter, r *http.Request, rr io.ReadCloser) {
+	p.logger.Log("Incoming request host", r.Host)
+	if p.AuthUser != "" && p.AuthPass != "" {
+		user, pass, ok := parseBasicProxyAuth(r.Header.Get("Proxy-Authorization"))
+		if !ok || user != p.AuthUser || pass != p.AuthPass {
+			//p.Logger.Warn("Authorization attempt with invalid credentials")
+			http.Error(w, http.StatusText(http.StatusProxyAuthRequired), http.StatusProxyAuthRequired)
+			return
+		}
+	}
+	if r.URL.Scheme == "http" {
+		p.ForwardingHTTPProxy.ServeHTTP(w, r)
+	} else {
+		p.handleTunneling(w, r, rr)
+	}
+}
+
+func (p *ForwardingProxy) handleTunneling(w http.ResponseWriter, r *http.Request, rr io.ReadCloser) {
+	if r.Method != http.MethodConnect {
+		//p.Logger.Info("Method not allowed", zap.String("method", r.Method))
+		http.Error(w, http.StatusText(http.StatusMethodNotAllowed), http.StatusMethodNotAllowed)
+		return
+	}
+
+	//p.Logger.Debug("Connecting", zap.String("host", r.Host))
+
+	destConn, err := net.DialTimeout("tcp", r.Host, p.DestDialTimeout)
+	if err != nil {
+		//p.Logger.Error("Destination dial failed", zap.Error(err))
+		http.Error(w, err.Error(), http.StatusServiceUnavailable)
+		return
+	}
+
+	now := time.Now()
+	destConn.SetReadDeadline(now.Add(p.DestReadTimeout))
+	destConn.SetWriteDeadline(now.Add(p.DestWriteTimeout))
+
+	//p.Logger.Debug("Connected", zap.String("host", r.Host))
+	fmt.Println("Connected to host", r.Host)
+
+	resp := http.Response{}
+	resp.StatusCode = 200
+	resp.Proto = "HTTP/1.1"
+	resp.ProtoMajor = 1
+	resp.ProtoMinor = 1
+
+	resp.Write(w)
+
+	w.(http.Flusher).Flush()
+
+	fmt.Println("Flushed to host", r.Host)
+
+	done := make(chan struct{})
+	go func() {
+		transfer(flushWriter{w}, destConn, log.NewContext(p.logger).With(
+			// "dst", msg.ForwardedHost,
+			"src", r.Host,
+		))
+		close(done)
+	}()
+
+	transfer(destConn, rr, log.NewContext(p.logger).With(
```

Comment:
Is the order of these two transfer calls important? Since as its coded I'm not sure it is predictable. I mean logically it feels like this call will happen first, but it might not. I don't know enough about go routines to be sure of that though.

Reply:
The same exact technique is used in tcpproxy

Reply:
okay